File: LEETCODE/LEETCODE-CPP/33.cpp
//33. Search in Rotated Sorted Array

class Solution {
public:
    int search(vector<int>& nums, int target) 
    {
        int mid, start=0, end=nums.size()-1;
        while(start <= end)
        {
            
            mid= start+(end-start)/2;
          
            if(nums[mid]==target)
                return mid;
            
            if(nums[mid] > nums[end])
            {
                if(nums[mid] > target && nums[start] <= target)
                     end=mid-1;
                else
                    start=mid+1;
            }
            
            else if(nums[mid] < nums[start])
            {
                if(nums[mid] < target && nums[end] >= target)
                    start=mid+1;
                else
                    end=mid-1;
            }
            
            else
            {
                if(nums[mid]<target)
                    start=mid+1;
                else
                    end=mid-1;
            }
        }
        return -1;
    }
};

File: LEETCODE/LEETCODE-JAVA/33.java
//33. Search in Rotated Sorted Array
//https://leetcode.com/problems/search-in-rotated-sorted-array/

//Methos-1 Binary search
class Solution {
    public int search(int[] nums, int target) {
        int start=0, end=nums.length-1;

        while(start<=end){
            int mid=start+(end-start)/2;

            if(nums[mid]==target){
                return mid;
            }
            
            if (nums[start] <= nums[mid]) {
                if (nums[start] <= target && nums[mid] >= target) {
                    end = mid - 1;
                } else {
                    start = mid + 1;
                }
            } else {
                if (nums[end] >= target && nums[mid] <= target) {
                    start = mid + 1;
                } else {
                    end = mid - 1;
                }
            }
        }

        return -1;
    }
}